File: library/src/main/java/com/lee/fileselector/FileSelector.java
package com.lee.fileselector;

import android.app.Activity;
import android.content.Intent;
import android.graphics.Color;

import androidx.annotation.ColorInt;
import androidx.annotation.ColorRes;
import androidx.core.content.ContextCompat;

import java.io.File;
import java.io.FileFilter;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

/**
 * @author Lee
 */
public class FileSelector {

    private static volatile FileSelector singleton;

    public String[] mFileTypes;
    public int mSortType;
    public int maxCount = 9;
    public int requestCode;
    public int barColor = Color.parseColor("#1bbc9b");
    public boolean isShow;
    public String[] selectPaths;
    public String[] ignorePaths;
    public Activity mActivity;
    public List<FileModel> mFileModels = Collections.synchronizedList(new ArrayList<FileModel>());
    public OnResultListener<FileModel> listener;
    public boolean isForEachStop;

    public FileSelector(Activity activity) {
        mActivity = activity;
        selectPaths = new String[]{"/storage/emulated/0/DCIM", "/storage/emulated/0/Android/data/" + mActivity.getPackageName() + "/"};
        mFileModels.clear();
    }

    public static FileSelector getInstance(Activity activity) {
        if (singleton == null) {
            synchronized (FileSelector.class) {
                if (singleton == null) {
                    singleton = new FileSelector(activity);
                }
            }
        }
        return singleton;
    }

    public FileSelector setSelectPath(String... selectPaths) {
        this.selectPaths = selectPaths;
        return this;
    }

    public FileSelector setIgnorePath(String... ignorePaths) {
        this.ignorePaths = ignorePaths;
        return this;
    }

    public FileSelector isShowHiddenFile(boolean isShow) {
        this.isShow = isShow;
        return this;
    }

    public FileSelector setBarColorRes(@ColorRes int barColor) {
        return this.setBarColorInt(ContextCompat.getColor(mActivity, barColor));
    }

    public FileSelector setBarColorInt(@ColorInt int barColor) {
        this.barColor = barColor;
        return this;
    }

    public FileSelector setMaxCount(int maxCount) {
        this.maxCount = Math.max(maxCount, 1);
        return this;
    }

    public FileSelector setFileType(String... fileTypes) {
        this.mFileTypes = fileTypes;
        return this;
    }

    public FileSelector setFileTypes(String[] fileTypes) {
        this.mFileTypes = fileTypes;
        return this;
    }

    public FileSelector setSortType(int sortType) {
        this.mSortType = sortType;
        return this;
    }

    public FileSelector requestCode(int requestCode) {
        this.requestCode = requestCode;
        return this;
    }

    public FileSelector forResult(OnResultListener<FileModel> listener) {
        this.listener = listener;
        return this;
    }

    public FileSelector setThreadStop(boolean isForEachStop) {
        this.isForEachStop = isForEachStop;
        return this;
    }

    public void start() {
        Intent intent = new Intent(mActivity, FileSelectorActivity.class);
        mActivity.startActivityForResult(intent, requestCode);
    }

    public List<FileModel> getFiles() {
        if (mFileFilter == null) {
            mFileFilter = new MyFileFilter(mFileTypes, isShow);
        }
        if (selectPaths != null && selectPaths.length > 0) {
            for (String selectPath : selectPaths) {
                //线程池开启任务
                GlobalThreadPools.getInstance().execute(new Runnable() {
                    @Override
                    public void run() {
                        getFolderFiles(selectPath);
                    }
                });
            }
        }
        try {
            Thread.sleep(100);
            while (!GlobalThreadPools.getInstance().hasDone()) {
                Thread.sleep(40);
            }
        } catch (InterruptedException e) {
            e.printStackTrace();
        }


        return mFileModels;
    }

    private FileFilter mFileFilter = null;

    public void getFolderFiles(String path) {
        if (!isForEachStop) {
            File file = new File(path);
            File[] files = file.listFiles(mFileFilter);
            if (files == null || files.length == 0) {
                return;
            }
            List<File> dirs = new ArrayList();
            List<FileModel> fms = new ArrayList();
            for (File value : files) {
                if (value.isFile()) {
                    FileModel fileModel = new FileModel();
                    String pathStr = value.getAbsolutePath();
                    fileModel.setExtension(FileUtils.getExtension(pathStr));
                    fileModel.setName(FileUtils.getName(pathStr));
                    fileModel.setPath(pathStr);
                    fileModel.setSize(value.length());
                    fileModel.setDate(value.lastModified());
                    fms.add(fileModel);
                } else {
                    //加入目录
                    dirs.add(value);
                }
            }
            mFileModels.addAll(fms);
            for (File dir : dirs) {
                if (ignorePaths != null && ignorePaths.length > 0) {
                    for (String ignorePath : ignorePaths) {
                        if (!(dir.getAbsolutePath().toLowerCase().contains(ignorePath.toLowerCase()))) {
                            //线程池开启任务
                            GlobalThreadPools.getInstance().execute(new Runnable() {
                                @Override
                                public void run() {
                                    getFolderFiles(dir.getAbsolutePath());
                                }
                            });
                        }
                    }
                } else {
                    //线程池开启任务
                    GlobalThreadPools.getInstance().execute(new Runnable() {
                        @Override
                        public void run() {
                            getFolderFiles(dir.getAbsolutePath());
                        }
                    });
                }
            }
        }
    }




}

File: library/src/main/kotlin/xyz/junerver/fileselector/FileSelector.kt
package xyz.junerver.fileselector

import android.content.Context
import android.graphics.Color
import android.os.Build
import android.os.Environment
import androidx.annotation.ColorInt
import androidx.annotation.ColorRes
import androidx.core.content.ContextCompat
import xyz.junerver.fileselector.worker.ActivityUIWorker
import xyz.junerver.fileselector.worker.FilesScanWorker
import java.lang.ref.SoftReference

/**
 * Description: 入口 & 配置
 * @author Junerver
 * date: 2021/12/27-9:50
 * Email: junerver@gmail.com
 * Version: v1.0
 */

const val VERSION = "0.0.15"

class FileSelector private constructor(ctx: Context) {
    //使用软引用持有ctx对象以避免内存泄漏
    private val mSrCtx: SoftReference<Context> = SoftReference(ctx)

    init {
        selectPaths = arrayOf(
            "/storage/emulated/0/DCIM",
            "/storage/emulated/0/Android/data/" + mSrCtx.get()?.packageName + "/"
        )
    }

    fun setSelectPath(vararg selects: String): FileSelector {
        selectPaths = selects as Array<String>
        if (!selectPaths.isNullOrEmpty()) {
            "~~~~~~~~~~~~~~~~~~~ 选择的文件目录 ~~~~~~~~~~~~~~~~~~~".log()
            selectPaths.forEach {
                it.log()
            }
        }
        return this
    }

    fun setIgnorePath(vararg paths: String): FileSelector {
        ignorePaths = paths as Array<String>
        if (!ignorePaths.isNullOrEmpty()) {
            "~~~~~~~~~~~~~~~~~~~ 忽略的文件目录 ~~~~~~~~~~~~~~~~~~~".log()
            ignorePaths.forEach {
                it.log()
            }
        }

        return this
    }

    fun isShowHiddenFile(show: Boolean): FileSelector {
        isShowHidden = show
        return this
    }

    fun setBarColorRes(@ColorRes barColor: Int): FileSelector {
        return setBarColorInt(ContextCompat.getColor(mSrCtx.get()!!, barColor))
    }

    fun setBarColorInt(@ColorInt color: Int): FileSelector {
        barColor = color
        return this
    }

    fun setMaxCount(max: Int): FileSelector {
        maxCount = Math.max(max, 1)
        return this
    }

    fun setFileType(vararg fileTypes: String): FileSelector {
        mFileTypes = fileTypes as Array<String>
        return this
    }

    fun setFileTypes(fileTypes: Array<String>?): FileSelector {
        if (fileTypes != null) {
            mFileTypes = fileTypes
        }
        return this
    }

    fun setSortType(sortType: Int): FileSelector {
        mSortType = sortType
        return this
    }

    fun setDebug(debug: Boolean): FileSelector {
        isDebugLog = debug
        return this
    }

    /**
    * Description: 主动请求权限，需要在resume中处理判断是否授权
    * @author Junerver
    * @date: 2021/12/30-9:09
    * @Email: junerver@gmail.com
    * @Version: v1.0
    * @param
    * @return
    */
    fun requestManagerFiles() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R && !Environment.isExternalStorageManager()) {
            //没有文件管理权限去申请
            mSrCtx.get()?.apply {
                showManagerFileTips(
                    request = {
                        startActivity(it)
                    }
                )
            }
        }
    }

    /**
     * Description: 启动内置的UI用于显示扫描结果
     * @author Junerver
     * @date: 2021/12/27-12:37
     * @Email: junerver@gmail.com
     * @Version: v1.0
     * @param
     * @return
     */
    fun startUIWorker(): ActivityUIWorker {
        return ActivityUIWorker(mSrCtx)
    }

    /**
     * Description: 仅按照配置扫描文件
     * @author Junerver
     * @Email: junerver@gmail.com
     * @Version: v1.0
     * @param
     * @return
     */
    fun startScanWorker(): FilesScanWorker {
        return FilesScanWorker
    }

    companion object {
        const val BY_NAME_ASC = 0
        const val BY_NAME_DESC = 1
        const val BY_TIME_ASC = 2
        const val BY_TIME_DESC = 3
        const val BY_SIZE_ASC = 4
        const val BY_SIZE_DESC = 5
        const val BY_EXTENSION_ASC = 6
        const val BY_EXTENSION_DESC = 7

        internal var mFileTypes: Array<String> = arrayOf()
        internal var mSortType = BY_TIME_DESC
        internal var maxCount = 9
        internal var barColor = Color.parseColor("#1bbc9b")
        internal var isShowHidden = false
        internal var selectPaths: Array<String> = arrayOf()
        internal var ignorePaths: Array<String> = arrayOf()
        private var instance: FileSelector? = null
        internal var isDebugLog = false

        fun getInstance(context: Context): FileSelector {
            "++++++++++++++++++ Versin: $VERSION ++++++++++++++++++++++".log()
            if (instance == null) {
                instance = FileSelector(context)
            }
            return instance as FileSelector
        }
    }
}